File: marking_machine.nim
## marking_machine.nim

import os
import strutils
import streams
import times

import serial # Or: `import serial/utils`
import parsetoml

import lib/db_redis
import lib/logging
import lib/utils

let hApp = newHApp()
let config = hApp.config

proc get_ports():seq[string] = 

    var port_list:seq[string]
    
    for port in listSerialPorts():
        ##  echo port
        port_list.add(port)
          
        #info(port)
    return port_list  
    
proc main() = 

    info("start...")    

    let port_list = get_ports()
    
    info("RS232 Ports: " & port_list.join(sep=","))
    
    let INTERVAL = parsetoml.getInt(config["app"]["interval"])
    
    ## rs232
    let machine = config["machine"]
    
    let portName = parsetoml.getStr(machine["port"])
    let baud_rate = int32(parsetoml.getInt(machine["baudRate"]))
    let byteSize = byte(parsetoml.getInt(machine["byteSize"]))
    let parity = parsetoml.getInt(machine["parity"])
    let stopBits = parsetoml.getInt(machine["stopBits"])
    let timeout = int32(parsetoml.getInt(machine["timeout"]))
    
    let buffer_len = parsetoml.getInt(machine["buffer_len"])
    
    let port = newSerialPort(portName)
    
    ## redis
    let init_fn = parsetoml.getStr(config["redis"]["init_lua"],"init.lua")    
    var fs2 = newFileStream(init_fn, fmRead)    
    let init_script = fs2.readAll()
    
    let lua_fn = parsetoml.getStr(config["redis"]["enqueue_lua"],"enqueue.lua")    
    var fs = newFileStream(lua_fn, fmRead)    
    let lua_script = fs.readAll()    
    
    var sha1:string    
    
    let rdb = newRedisDB(config["redis"])
    if rdb.enable:
        let rtn = rdb.exec("EVAL", @[init_script, "0"])
        ##  # load lua script
        sha1 = rdb.exec("SCRIPT", @["LOAD", lua_script])     
    
    ## open Serial Port
    port.open(baud_rate, Parity(parity), byteSize, 
              StopBits(stopBits), readTimeout = timeout)
    
    var receiveBuffer = newString(buffer_len)
    
    var numReceived:int
    
    while true:
        #let rtn = port.write("abc")
        #echo rtn
        ##  createdon = format(now(),"yyyy-MM-dd'T'HH:mm:ss")
        try:
            ## read
            numReceived = port.read(receiveBuffer)
        except TimeoutError:
            echo "timeout"
            sleep(100)
            continue
        except:
            echo "error"
            sleep(100)
            continue
      
        let data_str = receiveBuffer[0 ..< numReceived]

        echo getClockStr() & "," &  receiveBuffer[0 ..< numReceived]

        let ts = epochTime().formatFloat(ffDecimal, 4)

        let eval_rtn = rdb.exec("EVALSHA", @[sha1, "1", "ws", ts, receiveBuffer[0 ..< numReceived]])
        echo eval_rtn
        ## write
        discard port.write(receiveBuffer[0 ..< numReceived])
        ## sleep in ms
        sleep(INTERVAL)
    
    #port.close()

when isMainModule:
    
    main()

File: marking_machine.py

import time
import json

import redis
import serial
import serial.tools.list_ports

REDIS_HOST = 'localhost'
REDIS_PORT = 6379

PORT = "COM8"
BAUDRATE = 9600
BYTESIZE = 8
PARITY = serial.PARITY_ODD
STOPBITS = 1

INTERVAL = 0.6

MSG = b"M1T X+12.68mm Y+15.79mm Z+25.68mm\r\n"

def main():
    
    plist = list(serial.tools.list_ports.comports())
    print("RS232 Ports: ", end="")
    for p in plist:
        print(p[0], end=",")
    print()

    ## redis client
    red = redis.StrictRedis(host=REDIS_HOST, port=REDIS_PORT, db=0)
    
    ## init script
    with open("conf/marking_machine_init.lua","r") as fh:
        lua_init = fh.read()    
    
    init_data = red.register_script(lua_init)
    r = init_data(keys=[], args=[])
    #print(y)
    
    ## enqueue script
    with open("conf/marking_machine_enqueue.lua","r") as fh2:
        lua = fh2.read()
    #print(lua)
    multiply = red.register_script(lua)
    
    ## open serial port
    port = serial.Serial(port=PORT, baudrate=BAUDRATE, bytesize=BYTESIZE, 
                         parity=PARITY, stopbits=STOPBITS, timeout=0)
    ## loop here
    while True:
        ## read msg from serial port
        msg = port.readall()
        
        if msg != b'':
            print(msg)            

            y = multiply(keys=['MTS01'], args=[time.time(), msg])
            
            print(y)
            
            port.write(MSG)
        
        else:
            print("no data")
            port.write(MSG)
            
        time.sleep(INTERVAL)
   
if __name__ == "__main__"    :
    
    main()